leaktest_test.go

```diff
@@ -19,68 +20,121 @@ func (tr *testReporter) Errorf(format string, args ...interface{}) {
+		name       string
+		expectLeak bool
+	}{
+		{
+			name:       "Infinite for loop",
+			expectLeak: true,
+			f: func() {
+				for {
+					time.Sleep(time.Second)
+				}
+			},
+		},
+		{
+			name:       "Select on a channel not referenced by other goroutines.",
+			expectLeak: true,
+			f: func() {
+				c := make(chan struct{})
+				<-c
+			},
+		},
+		{
+			name:       "Blocked select on channels not referenced by other goroutines.",
+			expectLeak: true,
+			f: func() {
+				c := make(chan struct{})
+				c2 := make(chan struct{})
+				select {
+				case <-c:
+				case c2 <- struct{}{}:
+				}
+			},
+		},
+		{
+			name:       "Blocking wait on sync.Mutex that isn't referenced by other goroutines.",
+			expectLeak: true,
+			f: func() {
+				var mu sync.Mutex
+				mu.Lock()
+				mu.Lock()
+			},
+		},
+		{
+			name:       "Blocking wait on sync.RWMutex that isn't referenced by other goroutines.",
+			expectLeak: true,
+			f: func() {
+				var mu sync.RWMutex
+				mu.RLock()
+				mu.Lock()
+			},
+		},
+		{
+			name:       "HTTP Client with KeepAlive Disabled.",
+			expectLeak: false,
+			f: func() {
+				tr := &http.Transport{
+					DisableKeepAlives: true,
+				}
+				client := &http.Client{Transport: tr}
+				_, err := client.Get("http://localhost:8091")
+				if err != nil {
+					t.Error(err)
+				}
+			},
+		},
+		{
+			name:       "HTTP Client with KeepAlive Enabled.",
+			expectLeak: true,
+			f: func() {
+				tr := &http.Transport{
+					DisableKeepAlives: false,
+				}
+				client := &http.Client{Transport: tr}
+				_, err := client.Get("http://localhost:8091")
+				if err != nil {
+					t.Error(err)
+				}
+			},
+		},
+	}
+
+	// Start our keep alive server for keep alive tests
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go startKeepAliveEnabledServer(ctx)
+
 	// this works because the running goroutine is left running at the
 	// start of the next test case - so the previous leaks don't affect the
 	// check for the next one
-	for i, fn := range leakyFuncs {
-		checker := &testReporter{}
-		snapshot := CheckTimeout(checker, time.Second)
-		go fn()
-
-		snapshot()
-		if !checker.failed {
-			t.Errorf("didn't catch sleeping goroutine, test #%d", i)
-		}
+	for _, leakyTestcase := range leakyFuncs {
+
+		t.Run(leakyTestcase.name, func(t *testing.T) {
+			checker := &testReporter{}
+			snapshot := CheckTimeout(checker, time.Second)
+			go leakyTestcase.f()
+
+			snapshot()
+
+			if !checker.failed && leakyTestcase.expectLeak {
+				t.Error("didn't catch sleeping goroutine")
+			}
+			if checker.failed && !leakyTestcase.expectLeak {
+				t.Error("got leak but didn't expect it")
+			}
+		})
 	}
 }
 
 // TestSlowTest verifies that the timeout works on slow tests: it should
 // be based on time after the test finishes rather than time after the test's
 // start.
 func TestSlowTest(t *testing.T) {
-	defer CheckTimeout(t, 1000 * time.Millisecond)()
+	defer CheckTimeout(t, 1000*time.Millisecond)()
```

Comment:
is this a gofmt change?

Reply:
Yes it looks like `gofmt` is picking that one up. 

```
gofmt -d leaktest_test.go 
diff -u leaktest_test.go.orig leaktest_test.go
--- leaktest_test.go.orig	2018-11-06 16:10:23.301798204 +0000
+++ leaktest_test.go	2018-11-06 16:10:23.305798204 +0000
@@ -138,7 +138,7 @@
 // be based on time after the test finishes rather than time after the test's
 // start.
 func TestSlowTest(t *testing.T) {
-	defer CheckTimeout(t, 1000 * time.Millisecond)()
+	defer CheckTimeout(t, 1000*time.Millisecond)()
 
 	go time.Sleep(1500 * time.Millisecond)
 	time.Sleep(750 * time.Millisecond)
```

Happy to roll back or leave, whichever you'd like.